{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio import Align\n",
    "import numpy as np\n",
    "import math\n",
    "##### \n",
    "#\n",
    "#\n",
    "#\n",
    "#\n",
    "\n",
    "\n",
    "\n",
    "# >>> muscle_exe = r\"C:\\Program Files\\Aligments\\muscle3.8.31_i86win32.exe\"\n",
    "# >>> in_file = r\"C:\\My Documents\\unaligned.fasta\"\n",
    "# >>> out_file = r\"C:\\My Documents\\aligned.fasta\"\n",
    "# >>> muscle_cline = MuscleCommandline(muscle_exe, input=in_file, out=out_file)\n",
    "# >>> print(muscle_cline)\n",
    "\n",
    "# fetch working directory from user\n",
    "file_path = input('Enter file path to current work directory: ')\n",
    "\n",
    "\n",
    "seq_read = SeqIO.read(file_path + 'data/data1.txt', 'fasta')\n",
    "ref_read = SeqIO.read(file_path + 'data/ref.txt', 'fasta')\n",
    "\n",
    "aligner = Align.PairwiseAligner()\n",
    "\n",
    "aligner.mode = 'local'\n",
    "\n",
    "seq1 = seq_read.seq\n",
    "seq2 = ref_read.seq\n",
    "\n",
    "score = aligner.score(seq1, seq2)\n",
    "print(score)\n",
    "\n",
    "alignments = aligner.align(seq1, seq2)\n",
    "for alignment in alignments:\n",
    "    print(alignment)\n",
    "\n",
    "\n",
    "# fetch reference file name from user\n",
    "ref_name = input('Enter reference file name: ')\n",
    "\n",
    "\n",
    "## the below code calculates the k2p distance matrix from FASTA files\n",
    "seqs = []\n",
    "my_array = []\n",
    "for i in range(1, 10):\n",
    "\tfor record in SeqIO.parse('data' + str(i)+ '.txt', 'fasta'):\n",
    "\t\tseqs.append(record.seq)\n",
    "\n",
    "\n",
    "#calculate proportion of transitions\n",
    "def p_score(seq1, seq2, cap):\n",
    "\tp_count = 0\n",
    "\t\tfor i in range(0, cap):\n",
    "\t\t\tif (seq1[i] == 'a' and seq2[i] == 'g') or (seq1[i] == 'g' and seq2[i] == 'a') or (seq1[i] == 't' and seq2[i] == 'c') or (seq1[i] == 'c' and seq2[i] == 't'):\n",
    "\t\t\t\tp_count += 1\n",
    "\t\t\t\t\tprint(p_count, cap, p_count / cap)\n",
    "\t\t\t\t\treturn (p_count / cap)\n",
    "\n",
    "\n",
    "#calculate proportion of transversions\n",
    "def q_score(seq1, seq2, cap):\n",
    "\tq_count = 0\n",
    "\t\tfor j in range(0, cap):\n",
    "\t\t\tif (seq1[j] == 'a' and seq2[j] == 'c') or (seq1[j] == 'c' and seq2[j] == 'a') or (seq1[j] == 'c' and seq2[j] == 'g') or (seq1[j] == 'g' and seq2[j] == 'c') or (seq1[j] == 'a' and seq2[j] == 't') or (seq1[j] == 't' and seq2[j] == 'a') or (seq1[j] == 'g' and seq2[j] == 't') or (seq1[j] == 't' and seq2[j] == 'g'):\n",
    "\t\t\t\tq_count += 1\n",
    "\t\t\t\t\tprint(q_count, cap, q_count / cap)\n",
    "\t\t\t\t\treturn (q_count / cap)\n",
    "\n",
    "\n",
    "#calculate final k2p score\n",
    "def k2p(seq1, seq2):\n",
    "\tif len(seq1) < len(seq2): #cap length of comparison to shorter of two seqs\n",
    "\t\tcap = len(seq1)\n",
    "\t\t\telse:\n",
    "\t\t\t\tcap = len(seq2)\n",
    "\t\t\t\t\t\n",
    "\t\t\t\t\tp = p_score(seq1, seq2, cap)\n",
    "\t\t\t\t\tq = q_score(seq1, seq2, cap)\n",
    "\t\t\t\t\tscore = -0.5 * np.log((1 - 2*p - q) * math.sqrt(1 - 2*q))\n",
    "\t\t\t\t\t#print(p, q, score)\n",
    "\t\t\t\t\t\treturn (-0.5 * np.log((1 - 2*p - q) * math.sqrt(1 - 2*q)))\n",
    "\n",
    "\n",
    "#slot k2p values into distance matrix array\n",
    "def dist_mat(seqs):\n",
    "\tmy_array = np.empty([9, 9])\n",
    "\tfor i in range(len(seqs)):\n",
    "\t\tfor j in range(len(seqs)):\n",
    "\t\t\tmy_array[i,j] = k2p(seqs[i].lower(), seqs[j].lower())\n",
    "\t\t\tmy_array[i,j] = round(my_array[i,j],4)\n",
    "\t\t\tnp.savetxt('tablefile.txt', my_array, fmt='%.4f')\n",
    "\tprint(my_array)\n",
    "\treturn my_array\n",
    "\n",
    "dist_mat(seqs)\n",
    "#\n",
    "# /Users/madelyne/Desktop/mtgenome-pipeline/\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
